### DISCUSSION TITLE
Optional Redis master

### DISCUSSION THREAD

**Discussion by AlexWayfer** [anycable/anycable#136] [Q&A] [answered]
Hello.

We have a Kubernetes setup with Redis sentinels, but without specific master.

From [the Redis documentation](https://redis.io/topics/sentinel):

> The advantage of having multiple Sentinel processes cooperating are the following:
>
> 1. Failure detection is performed when multiple Sentinels agree about the fact a given master is no longer available. This lowers the probability of false positives.

So, master can be unavailable by design. And we're not sure what sentinel is current master.

But, as I see and judging by recieved logs and [`anycable-helm` template](https://github.com/anycable/anycable-helm/blob/f04cc3b/anycable-go/templates/env-secret.yml#L3), `ANYCABLE_REDIS_URL` for Redis master is kind of required.

```
2021-07-16 17:15:07 | I 2021-07-16T14:15:07.586Z context=main Starting AnyCable 1.0.2-ce5b4c5 (pid: 1, open file limit: 1048576)
-- | --
  |   | 2021-07-16 17:15:07 | I 2021-07-16T14:15:07.586Z context=main Handle WebSocket connections at http://0.0.0.0:8080/cable
  |   | 2021-07-16 17:15:07 | I 2021-07-16T14:15:07.586Z context=main Handle health connections at http://0.0.0.0:8080/health
  |   | 2021-07-16 17:15:07 | I 2021-07-16T14:15:07.587Z context=metrics Serve metrics at http://0.0.0.0:8081/metrics
  |   | 2021-07-16 17:15:07 | I 2021-07-16T14:15:07.587Z context=rpc RPC controller initialized: localhost:50051 (concurrency: 28, proto_versions: v0,v1)
  |   | 2021-07-16 17:15:07 | W 2021-07-16T14:15:07.687Z context=pubsub Failed to get master address from sentinel.
  |   | 2021-07-16 17:15:07 | W 2021-07-16T14:15:07.687Z context=pubsub Failed to discover sentinels
  |   | 2021-07-16 17:15:07 | E 2021-07-16T14:15:07.687Z context=main !!! Subscriber failed !!!
  |   | 2021-07-16 17:15:07 | redigo: no sentinels available; last error: redigo: nil returned
```

Can be Redis master optional and can you tell us how to configure properly for such case?

**Comment by Envek** [comment, marked as answer]
Hey! Look at description of https://github.com/anycable/anycable-helm/pull/9:

> Example of deploying anycable-go with sentinels:
>
> ```sh
> helm install anycable-redis bitnami/redis --set "cluster.enabled=true,sentinel.enabled=true,password=password"
> helm install anycable-go anycable/anycable-go --set "env.anycableRedisUrl=redis://:password@mymaster:6379/0,env.anycableRedisSentinels=:password@anycable-redis-headless:26379"
> ```
>
> Note that in this example in `ANYCABLE_REDIS_URL` _name of the master instance_ is specified and in `ANYCABLE_REDIS_SENTINELS` address of headless service is specified (which points into all redis pods: both initial master and slaves)

When sentinels are used, `ANYCABLE_REDIS_URL` must contain _name_ of Redis deployment in sentinel configuration as Sentinels can keep track of multiple Redis deployments with different names.

**Reply by romcheck** [reply]
Hey!
We have the same problem.
What do you mean by 'name of Redis deployment in sentinel configuration'?
We do not have any redis deployment. We use managed redis in some cloud provider and have only sentinel hosts (list of names or ip addresses)

**Reply by Envek** [reply]
In [Redis Sentinel docs](https://redis.io/topics/sentinel#configuring-sentinel) it is called `master-group-name` or _name of set of Redis instances_. Your cloud provider should also provide this name along with sentinel hosts (by default it can be called `mymaster` as in examples in the official Redis docs).

**Reply by romcheck** [reply]
Thanks a lot. It is working now.
